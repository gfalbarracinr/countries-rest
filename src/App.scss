@use "index.scss" as variables;
body{
  background-color: variables.$lightGray;
  margin: 0;
}
.container{
  margin: 0;
  height: 100%;
  min-height: 100vh;
}
.country-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  padding: variables.$upperTopPadding variables.$sideTopPadding;
}
.container-dark{
  background-color: variables.$veryDarkBlue;
}