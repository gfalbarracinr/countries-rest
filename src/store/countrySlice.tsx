/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';
import Action from '../states/Action';
import CountryState from '../states/CountryState';
import countryFilterType from '../types/CountryFilterType';

const initialState: CountryState = {
  countryFilterByRegion: 'all',
  countryGlobalSearch: '',
};
const countrySlice = createSlice({
  name: 'countries',
  initialState,
  reducers: {
    changeFilter(state: CountryState, action: Action<countryFilterType>) {
      state.countryFilterByRegion = action.payload;
    },

    findingByName(state: CountryState, action: Action<string>) {
      state.countryGlobalSearch = action.payload;
    },
  },
});

export const { actions } = countrySlice;
export default countrySlice;
