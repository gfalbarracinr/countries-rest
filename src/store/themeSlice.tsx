/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';
import ThemeState from '../states/ThemeState';
import Mode from '../types/Mode';

const initialState: ThemeState = {
  mode: Mode.Light,
};
const themeSlice = createSlice({
  name: 'Theme',
  initialState,
  reducers: {
    toggleTheme(state: ThemeState) {
      if (state.mode === Mode.Light) {
        state.mode = Mode.Dark;
      } else {
        state.mode = Mode.Light;
      }
    },
  },

});

export const { actions } = themeSlice;

export default themeSlice;
