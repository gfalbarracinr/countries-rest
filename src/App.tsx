import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import './App.scss';
import Header from './components/header';
import Country from './types/Country';
import CountryInfo from './components/country';
import Nav from './components/nav';
import countryFilterType from './types/CountryFilterType';
import { RootState } from './store';
import Mode from './types/Mode';
import Loading from './components/loading';
import Error from './components/error';

function App() {
  const [countries, setCountries] = useState<Array<Country>>([]);
  const [error, setError] = useState<string>('');
  const [loading, setLoading] = useState<boolean>(false);
  const [currentCountries, setCurrentCountries] = useState<Array<Country>>([]);
  const countryFilterByRegion: countryFilterType = useSelector(
    (state: RootState) => state.country.countryFilterByRegion,
  );
  const countryGlobalSearch: string = useSelector(
    (state: RootState) => state.country.countryGlobalSearch,
  );

  const theme: Mode = useSelector(
    (state: RootState) => state.theme.mode,
  );

  const filterCountries = (data: Array<any>) => data.map((
    country:any,
  ) => {
    const capital: string = country.capital ? country.capital[0] : '';
    const land: Country = {
      name: country.name.common,
      population: country.population,
      region: country.region,
      capital,
      flag: country.flags.png,
      code: country.altSpellings[0],
    };
    return land;
  });
  const fetchData = () => {
    setLoading(true);
    fetch('https://restcountries.com/v3.1/all')
      .then((response) => response.json())
      .then((data) => {
        const filteredCountries = filterCountries(data);
        setCountries(filteredCountries);
        setCurrentCountries(filteredCountries);
        setLoading(false);
        setError('');
      })
      .catch((e) => {
        setError(e.message);
        setLoading(false);
      });
  };
  useEffect(() => {
    fetchData();
  }, []);
  const filterCountry = (elements: Array<Country>) => {
    if (countryFilterByRegion === 'all') {
      return elements;
    }
    return elements.filter(
      (country) => country.region === countryFilterByRegion,
    );
  };

  useEffect(() => {
    const filteredByRegion = filterCountry(countries);
    if (countryGlobalSearch === '') {
      setCurrentCountries(filteredByRegion);
    }
    const filterByName = filteredByRegion.filter(
      (country) => country.name.toLowerCase().startsWith(countryGlobalSearch.toLowerCase()),
    );
    setCurrentCountries(filterByName);
  }, [countryFilterByRegion, countryGlobalSearch]);
  return (
    <div className={`container container-${theme.toString()}`}>
      <Header />
      <Nav />
      { loading && <Loading />}
      { error !== '' && !loading && <Error message={error} />}
      <main className="country-container">
        { error === '' && !loading && currentCountries.map((country: Country) => (
          <CountryInfo
            key={country.name}
            name={country.name}
            population={country.population}
            region={country.region}
            capital={country.capital}
            flag={country.flag}
            code={country.code}
          />
        )) }
      </main>
    </div>
  );
}

export default App;
