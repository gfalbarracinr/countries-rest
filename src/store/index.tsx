import { configureStore } from '@reduxjs/toolkit';
import countrySlice from './countrySlice';
import themeSlice from './themeSlice';

const store = configureStore({
  reducer: {
    country: countrySlice.reducer,
    theme: themeSlice.reducer,
  },
});

export type RootState = ReturnType<typeof store.getState>
export default store;
