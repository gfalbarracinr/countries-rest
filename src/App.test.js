/* eslint-disable no-undef */
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import App from './App';
import store from './store';

describe('App component tests', () => {
  test('App snapshot test', () => {
    const components = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const tree = components.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Provider store={store}>
        <App />
      </Provider>,
      div,
    );
  });
});
