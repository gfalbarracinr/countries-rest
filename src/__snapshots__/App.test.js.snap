// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests App snapshot test 1`] = `
<div
  className="container container-light"
>
  <header
    className="header header-light"
  >
    <h1
      className="header-title"
    >
      Where in the world
    </h1>
    <button
      className="button button-light"
      onClick={[Function]}
      type="button"
    >
      Dark mode
    </button>
  </header>
  <nav
    className="navigation"
  >
    <input
      className="nav-search nav-search-light"
      onChange={[Function]}
      placeholder="Search for a country..."
      type="search"
    />
    <select
      className="nav-filter nav-filter-light"
      onChange={[Function]}
      value="Select a region"
    >
      <option
        className="nav-filter-element"
        value="all"
      >
        All regions
      </option>
      <option
        className="nav-filter-element"
        value="Europe"
      >
        Europe
      </option>
      <option
        className="nav-filter-element"
        value="Americas"
      >
        Americas
      </option>
      <option
        className="nav-filter-element"
        value="Asia"
      >
        Asia
      </option>
      <option
        className="nav-filter-element"
        value="Africa"
      >
        Africa
      </option>
      <option
        className="nav-filter-element"
        value="Oceania"
      >
        Oceania
      </option>
      <option
        className="nav-filter-element"
        value="Antarctic"
      >
        Antarctic
      </option>
    </select>
  </nav>
  <main
    className="country-container"
  />
</div>
`;

exports[`App snapshot test 1`] = `
<div
  className="container container-light"
>
  <header
    className="header header-light"
  >
    <h1
      className="header-title"
    >
      Where in the world
    </h1>
    <button
      className="button button-light"
      onClick={[Function]}
      type="button"
    >
      Dark mode
    </button>
  </header>
  <nav
    className="navigation"
  >
    <input
      className="nav-search nav-search-light"
      onChange={[Function]}
      placeholder="Search for a country..."
      type="search"
    />
    <select
      className="nav-filter nav-filter-light"
      onChange={[Function]}
      value="Select a region"
    >
      <option
        className="nav-filter-element"
        value="all"
      >
        All regions
      </option>
      <option
        className="nav-filter-element"
        value="Europe"
      >
        Europe
      </option>
      <option
        className="nav-filter-element"
        value="Americas"
      >
        Americas
      </option>
      <option
        className="nav-filter-element"
        value="Asia"
      >
        Asia
      </option>
      <option
        className="nav-filter-element"
        value="Africa"
      >
        Africa
      </option>
      <option
        className="nav-filter-element"
        value="Oceania"
      >
        Oceania
      </option>
      <option
        className="nav-filter-element"
        value="Antarctic"
      >
        Antarctic
      </option>
    </select>
  </nav>
  <main
    className="country-container"
  />
</div>
`;
